lte/gateway/python/integ_tests/s1aptests/test_enb_partial_reset_multi_ue_with_mme_restart.py

```diff
@@ -0,0 +1,122 @@
+        )
+
+    def tearDown(self):
+        self._s1ap_wrapper.cleanup()
+
+    def test_enb_partial_reset_multi_ue_with_mme_restart(self):
+        """ ENB Partial Reset with MME restart for 32 UEs """
+        ue_ids = []
+        num_ues = 32
+        self._s1ap_wrapper.configUEDevice(num_ues)
+        for _ in range(num_ues):
+            req = self._s1ap_wrapper.ue_req
+            print(
+                "************************* Calling attach for UE id ",
+                req.ue_id,
+            )
+            self._s1ap_wrapper.s1_util.attach(
+                req.ue_id,
+                s1ap_types.tfwCmd.UE_END_TO_END_ATTACH_REQUEST,
+                s1ap_types.tfwCmd.UE_ATTACH_ACCEPT_IND,
+                s1ap_types.ueAttachAccept_t,
+            )
+            ue_ids.append(req.ue_id)
+            # Wait on EMM Information from MME
+            self._s1ap_wrapper._s1_util.receive_emm_info()
+
+        # Add delay to ensure S1APTester sends attach partial before sending
+        # eNB Reset Request
+        time.sleep(0.5)
+
+        # Set the reset UEs list
+        random.seed(time.clock())
+        reset_ue_count = random.randint(1, num_ues)
+        random.seed(time.clock())
+        reset_ue_list = random.sample(range(num_ues), reset_ue_count)
+
+        print(
+            "************************* Sending eNB Partial Reset Request for",
+            reset_ue_count,
+            "UEs",
+        )
+        reset_req = s1ap_types.ResetReq()
+        reset_req.rstType = s1ap_types.resetType.PARTIAL_RESET.value
+        reset_req.cause = s1ap_types.ResetCause()
+        reset_req.cause.causeType = (
+            s1ap_types.NasNonDelCauseType.TFW_CAUSE_MISC.value
+        )
+        # Set the cause to MISC.hardware-failure
+        reset_req.cause.causeVal = 3
+        reset_req.r = s1ap_types.R()
+        reset_req.r.partialRst = s1ap_types.PartialReset()
+        reset_req.r.partialRst.numOfConn = reset_ue_count
+        reset_req.r.partialRst.ueS1apIdPairList = (
+            (s1ap_types.UeS1apIdPair) * reset_req.r.partialRst.numOfConn
+        )()
+        for indx in range(reset_req.r.partialRst.numOfConn):
+            reset_req.r.partialRst.ueS1apIdPairList[indx].ueId = ue_ids[
+                reset_ue_list[indx]
+            ]
+            print(
+                "Reset_req.r.partialRst.ueS1apIdPairList[indx].ueId",
+                reset_req.r.partialRst.ueS1apIdPairList[indx].ueId,
+                indx,
+            )
+
+        # Send eNB Partial Reset
+        self._s1ap_wrapper.s1_util.issue_cmd(
+            s1ap_types.tfwCmd.RESET_REQ, reset_req
+        )
+
+        print("************************* Restarting MME service on gateway")
+        self._s1ap_wrapper.magmad_util.restart_services(["mme"])
+
+        for j in range(30):
+            print("Waiting for", j, "seconds")
+            time.sleep(1)
+
+        response = self._s1ap_wrapper.s1_util.get_response()
```

Comment:
Is the "Reset_Ack" received after the MME restarts, or is it received by S1ap-tester before the restart itself, only the test case is reading it after the restart is complete?

A better check for state persistence after Partial Reset would be to confirm that S1 context for each UE was actually released. One way to do that would be to send a Service Request for each of the UEs which were reset.

Reply:
Reset Ack is received before MME restarts and the test case was reading it after 30 seconds of issuing restart command.

Added sending of service request for all the UE Ids which were reset. Also removed the sleep of 3 seconds for stateless mode, because in this case I could see MME was getting sufficient time to clean the UE context as part of ENB reset.